src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+                crd_group, crd_version, namespace, plural, obj_name
+            )
+        except ApiException as e:
+            if e.status == 404:
+                return
+            raise KubernetesError.from_exception(
+                "Error deleting gafaelfawringress",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+
+    async def get_observed_fileserver_state(
+        self, namespace: str
+    ) -> dict[str, bool]:
+        """Reconstruct the fileserver user map with what we can determine
+        from the Kubernetes cluster.
+
+        Objects with the name <username>-fs are presumed to be fileserver
+        objects, where <username> can be assumed to be the name of the
+        owning user.
+
+        It returns a dict mapping strings to the value True, indicating those
+        users who currently have fileservers.
+        """
+        observed_state: dict[str, bool] = {}
+        # Create fileserver namespace if necessary.
+        try:
+            await self.api.read_namespace(namespace)
+        except ApiException as e:
+            if e.status == 404:
+                self._logger.warning(f"No fileserver namespace '{namespace}'")
+                return observed_state
+            else:
+                raise KubernetesError.from_exception(
+                    "Error reading namespace",
+                    e,
+                    namespace=namespace,
+                ) from e
+        # Get all jobs
+        all_jobs = await self.batch_api.list_namespaced_job(namespace)
+        # Filter to those with the right label
+        users = [
+            x.metadata.labels.get("lsst.io/user")
+            for x in all_jobs.items
+            if x.metadata.labels.get("lsst.io/category", "") == "fileserver"
+        ]
+        # For each of these, check whether the fileserver is present
+        for user in users:
+            self._logger.debug(f"Checking user {user}")
+            good = await self.check_fileserver_present(
+                username=user, namespace=namespace
+            )
+            self._logger.debug(f"{user} fileserver good?  {good}")
+            if good:
+                observed_state[user] = True
+        return observed_state
+
+    async def check_fileserver_present(
+        self, username: str, namespace: str
+    ) -> bool:
+        """Our determination of whether a user has a fileserver is this:
+
+        We assume all fileserver objects are named <username>-fs, which we
+        can do, since we created them and that's the convention we chose.
+
+        A fileserver is working if:
+
+        1) it has a Job with at least one active Pod
+        2) it has a Service (properties not checked)
+        3) it has an Ingress (properties not checked)
+
+        We do not check the GafaelfawrIngress, because the Custom API is
+        clumsy, and the created Ingress is a good proxy.
```

Comment:
Well, more to the point, it doesn't matter whether there is a `GafaelfawrIngress`; the fileserver will only work if the `Ingress` exists. If `gafaelfawr-operator` is failing, the fileserver will never come up from a user's perspective.

Reply:
Amended.